storage/innobase/include/arch0recv.h

```diff
@@ -0,0 +1,347 @@
+  Arch_Dblwr_Ctx(Arch_Dblwr_Ctx const &) = delete;
+
+  /** Disable assignment. */
+  Arch_Dblwr_Ctx &operator=(Arch_Dblwr_Ctx const &) = delete;
+
+ private:
+  /** Buffer to hold the contents of the doublwrite buffer. */
+  byte *m_buf{nullptr};
+
+  /** Total file size of the file which holds the doublewrite buffer. */
+  uint64_t m_file_size{};
+
+  /** Doublewrite buffer file context. */
+  Arch_File_Ctx m_file_ctx;
+
+  /** List of doublewrite buffer blocks. */
+  Arch_Dblwr_Blocks m_blocks{};
+};
+
+/** Recovery system data structure for the archiver. */
+class Arch_Page_Sys::Recovery {
+ public:
+  /** Constructor: Initialize members
+  @param[in,out]  page_sys  global dirty page archive system
+  @param[in]      dir_name  main archiver directory name */
+  Recovery(Arch_Page_Sys *page_sys, const char *dir_name)
+      : m_arch_dir_name(dir_name), m_page_sys(page_sys) {}
+
+  /** Destructor: Close open file and free resources */
+  ~Recovery() = default;
+
+  /** Initialise the archiver's recovery system.
+  @return error code. */
+  dberr_t init_dblwr();
+
+  /** Scan the archive directory and fetch all info related to group
+  directories and its files.
+  @return true if the scan was successful. */
+  bool scan_for_groups();
+
+#ifdef UNIV_DEBUG
+  /** Print information related to the archiver recovery system added
+  for debugging purposes. */
+  void print();
+#endif
+
+  /** Parse for group information and fill the group.
+  @return error code. */
+  dberr_t recover();
+
+  /** Load archiver with the related data and start tracking if required.
+  @return error code. */
+  dberr_t load_archiver();
+
+  /** Disable copy construction */
+  Recovery(Recovery const &) = delete;
+
+  /** Disable assignment */
+  Recovery &operator=(Recovery const &) = delete;
+
+ private:
+  /** Read all the group directories and store information related to them
+  required for parsing.
+  @param[in]    file_path       file path information */
+  void read_group_dirs(const std::string file_path);
+
+  /** Read all the archived files belonging to a group and store information
+  related to them required for parsing.
+  @param[in]    dir_path        dir path information
+  @param[in]    file_path       file path information */
+  void read_group_files(const std::string dir_path,
+                        const std::string file_path);
+
+ private:
+  /** Archive directory. */
+  std::string m_arch_dir_name;
+
+  /** Global dirty page archive system */
+  Arch_Page_Sys *m_page_sys;
+
+  /** Doublewrite buffer context. */
+  Arch_Dblwr_Ctx m_dblwr_ctx{};
+
+  /** Mapping of group directory names and group information related to
+  the group. */
+  Arch_Dir_Group_Info_Map m_dir_group_info_map{};
+};
+
+/** Recovery system data structure for the archiver. */
+class Arch_Group::Recovery {
+ public:
+  /** Constructor.
+  @param[in]  group the parent class group object */
+  Recovery(Arch_Group *group) {
+    ut_ad(group != nullptr);
+    m_group = group;
+  }
```

Comment:
Unreachable class. Why do we not use an initializer list in the constructor and a `nonnull` attribute?